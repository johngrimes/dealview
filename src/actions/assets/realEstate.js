// @flow

import { putObject, deleteObject, getAllObjects } from 'db/db'
import AssetActions from 'actions/assets/assets'
import { realEstateToAsset } from 'types/assets/realEstate'
import { createObjectActions } from 'actions/objects'
import type { RealEstate } from 'types/assets/realEstate'
import type { Thunk } from 'types/commonTypes'

const objectStore = 'Asset.RealEstate'

const actions = createObjectActions('REAL_ESTATE', 'REAL_ESTATE',
                                    'RealEstate', 'RealEstate')

const putRealEstate = (realEstate: RealEstate): Thunk => {
  return dispatch => {
    dispatch(actions.putRealEstateRequest(realEstate))
    return new Promise(async (resolve, reject) => {
      try {
        const saved = await putObject(objectStore, realEstate)
        dispatch(actions.putRealEstateSuccess(saved))
        await dispatch(AssetActions.putAsset(realEstateToAsset(saved)))
        resolve(saved)
      } catch (error) {
        dispatch(actions.putRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const deleteRealEstate = (id: string): Thunk => {
  return dispatch => {
    dispatch(actions.deleteRealEstateRequest(id))
    return new Promise(async (resolve, reject) => {
      try {
        const deletedId = await deleteObject(objectStore, id)
        dispatch(AssetActions.deleteAsset(deletedId))
        dispatch(actions.deleteRealEstateSuccess(deletedId))
        resolve(deletedId)
      } catch (error) {
        dispatch(actions.deleteRealEstateFailure(error))
        reject(error)
      }
    })
  }
}

const loadRealEstate = (): Thunk => {
  return dispatch => {
    dispatch(actions.loadRealEstateRequest())
    return new Promise((resolve, reject) => {
      getAllObjects(objectStore).then(assets => {
        dispatch(actions.loadRealEstateSuccess(assets))
        resolve(assets)
      }).catch(error => {
        dispatch(actions.loadRealEstateFailure(error))
        reject(error)
      })
    })
  }
}

export default {
  ...actions,
  putRealEstate,
  deleteRealEstate,
  loadRealEstate,
}
