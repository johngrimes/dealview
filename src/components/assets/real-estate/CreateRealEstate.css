.create-real-estate {
  padding: 0 16px;
}

.create-real-estate .breadcrumbs {
  margin: 0 -16px 16px;
}
