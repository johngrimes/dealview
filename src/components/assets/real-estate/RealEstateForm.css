.real-estate-form {
  padding: 0 16px;
}

.form-page-inactive {
  display: none;
}
