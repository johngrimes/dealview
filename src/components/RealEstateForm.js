import React from 'react'
import PropTypes from 'prop-types'
import _ from 'lodash'
import moment from 'moment'

import InputField from './InputField.js'
import CurrencyField from './CurrencyField.js'
import HiddenField from './HiddenField.js'
import AddressField, { AddressErrorsDefaults } from './AddressField.js'
import TextAreaField from './TextAreaField.js'
import ValuationsInput from './ValuationsInput.js'
import DateField from './DateField.js'
import * as Validations from '../utils/formValidation.js'
import { RealEstateDefaults } from '../data/realEstate.js'
import { DateStorageFormat } from '../data/commonTypes.js'
import {
  updateValuationsWithPurchaseOrSale,
  updateValuationsWithPurchase,
  updateValuationsWithSale,
  purchaseFilter,
  saleFilter,
} from '../data/valuations.js'

import '../styles/forms.css'
import '../styles/buttons.css'
import '../styles/tables.css'
import './styles/RealEstateForm.css'

const RealEstateErrorsDefaults = {
  name: [],
  address: AddressErrorsDefaults,
  notes: [],
  purchaseDate: [],
  purchaseAmount: [],
  saleDate: [],
  saleAmount: [],
  valuations: [],
}

class RealEstateForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      realEstate: RealEstateDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.realEstate) {
      const realEstate = props.realEstate
      const valuations = props.realEstate.valuations
        ? props.realEstate.valuations
        : []
      this.state = {
        ...this.state,
        realEstate,
        purchase: _.pick(valuations.find(purchaseFilter), [ 'date', 'amount' ]),
        sale: _.pick(valuations.find(saleFilter), [ 'date', 'amount' ]),
        errors: RealEstateForm.validate(realEstate),
        allErrorsShown: true,
      }
    }

    this.handleFocus = this.handleFocus.bind(this)
    this.handleAddressChange = this.handleAddressChange.bind(this)
    this.handleValuationsChange = this.handleValuationsChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName, value) {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, [fieldName]: value }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleFocus(fieldName) {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleAddressChange(address) {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, address }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleValuationsChange(valuations) {
    this.setState(prevState => {
      let purchaseVal = valuations.find(v => v.type === 'purchase')
      let saleVal = valuations.find(v => v.type === 'sale')
      let updatedValuations = valuations
      if (purchaseVal) {
        purchaseVal = _.pick(purchaseVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithPurchase(
          valuations,
          purchaseVal
        )
      }
      if (saleVal) {
        saleVal = _.pick(saleVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithSale(valuations, saleVal)
      }
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
      }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseOrSaleChange(purchaseOrSale, type) {
    this.setState(prevState => {
      const updatedPurchaseOrSale = { ...prevState[type], ...purchaseOrSale }
      const updatedValuations = updateValuationsWithPurchaseOrSale(
        prevState.realEstate.valuations,
        updatedPurchaseOrSale,
        type
      )
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
        [type === 'purchase' // eslint-disable-line
          ? 'purchaseDate'
          : 'saleDate']: updatedPurchaseOrSale.date,
      }
      return {
        ...prevState,
        realEstate,
        [type === 'purchase' ? 'purchase' : 'sale']: updatedPurchaseOrSale,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseChange(purchase) {
    return this.handlePurchaseOrSaleChange(purchase, 'purchase')
  }

  handleSaleChange(sale) {
    return this.handlePurchaseOrSaleChange(sale, 'sale')
  }

  handleSubmit(event) {
    event.preventDefault()
    this.setState(
      () => ({ errors: RealEstateForm.validate(this.state.realEstate) }),
      () => {
        const { errors, realEstate } = this.state
        if (
          typeof errors !== 'undefined' &&
          Validations.areErrorsPresent(errors)
        ) {
          const firstErrorFieldName = Validations.findFirstErrorFieldName(
            errors
          )
          firstErrorFieldName
            ? this.setState(() => ({
              allErrorsShown: true,
              focusedInput: firstErrorFieldName,
            }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) {
            this.props.onSubmit(realEstate)
          }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.realEstate) {
      this.setState(() => ({ realEstate: nextProps.realEstate }))
    }
  }

  render() {
    const {
      realEstate,
      purchase,
      sale,
      allErrorsShown,
      focusedInput,
    } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField =
      typeof realEstate.id === 'string'
        ? <HiddenField name='id' value={realEstate.id} />
        : null
    const minValuationDate =
      purchase && purchase.date
        ? moment(purchase.date, DateStorageFormat).add(1, 'd')
        : undefined
    const maxValuationDate =
      sale && sale.date
        ? moment(sale.date, DateStorageFormat).subtract(1, 'd')
        : undefined

    return (
      <form
        className='real-estate-form form form-aligned'
        onSubmit={this.handleSubmit}
      >
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField
            name='name'
            type='text'
            label='Name'
            value={realEstate.name}
            errors={errors.name}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <AddressField
            name='address'
            address={realEstate.address}
            errors={errors.address}
            onChange={this.handleAddressChange}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <TextAreaField
            name='notes'
            label='Notes'
            value={realEstate.notes}
            onChange={value => this.handleChange('notes', value)}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Purchase</legend>
          <DateField
            name='purchaseDate'
            label='Purchase date'
            value={purchase ? purchase.date : undefined}
            errors={errors.purchaseDate}
            forceErrorDisplay={allErrorsShown}
            maxDate={maxValuationDate}
            onChange={value => this.handlePurchaseChange({ date: value })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='purchaseAmount'
            label='Purchase amount'
            value={
              purchase && purchase.amount
                ? purchase.amount.toString()
                : undefined
            }
            errors={errors.purchaseAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value =>
              this.handlePurchaseChange({ amount: parseInt(value, 10) })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Sale</legend>
          <DateField
            name='saleDate'
            label='Sale date'
            value={sale ? sale.date : undefined}
            errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            minDate={minValuationDate}
            onChange={value => this.handleSaleChange({ date: value })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
          <CurrencyField
            name='saleAmount'
            label='Sale amount'
            value={sale && sale.amount ? sale.amount.toString() : undefined}
            errors={errors.saleAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value =>
              this.handleSaleChange({ amount: parseInt(value, 10) })}
            onFocus={this.handleFocus}
            focus={focusedInput}
          />
        </fieldset>
        <fieldset>
          <legend>Value</legend>
          <ValuationsInput
            name='valuations'
            valuations={realEstate.valuations}
            minDate={minValuationDate}
            maxDate={maxValuationDate}
            errors={errors.valuations}
            forceErrorDisplay={allErrorsShown}
            onChange={this.handleValuationsChange}
            onFocus={this.handleFocus}
          />
        </fieldset>
        <fieldset>
          <legend>Mortgage</legend>
          <button type='submit' className='button button-primary'>
            Submit
          </button>
        </fieldset>
      </form>
    )
  }

  static validate(realEstate) {
    const name = []
      .concat(Validations.required(realEstate.name))
      .concat(Validations.minLength(realEstate.name, 3))
      .concat(Validations.maxLength(realEstate.name, 100))
    const address = AddressField.validate(realEstate.address)
    const purchaseDate = [].concat(
      Validations.required(realEstate.purchaseDate)
    )
    const valuations = []
      .concat(ValuationsInput.validate(realEstate.valuations))
      .concat(RealEstateForm.purchaseValuationPresent(realEstate))
      .concat(RealEstateForm.valuationsWithinPurchaseAndSale(realEstate))

    return {
      ...RealEstateErrorsDefaults,
      name,
      address,
      purchaseDate,
      valuations,
    }
  }

  static purchaseValuationPresent(realEstate) {
    return Validations.validate(
      realEstate,
      re => re.valuations.some(valuation => valuation.type === 'purchase'),
      'There must be a purchase valuation'
    )
  }

  static valuationsWithinPurchaseAndSale(realEstate) {
    const purchaseVal = realEstate.valuations.find(v => v.type === 'purchase')
    const saleVal = realEstate.valuations.find(v => v.type === 'sale')
    if (!purchaseVal) {
      return []
    }

    return Validations.validate(
      realEstate,
      re =>
        re.valuations.every(
          valuation =>
            [ 'purchase', 'sale' ].includes(valuation.type) ||
            (saleVal
              ? valuation.date > purchaseVal.date &&
                valuation.date < saleVal.date
              : valuation.date > purchaseVal.date)
        ),
      'Valuations must all be between the purchase and sale dates'
    )
  }
}

RealEstateForm.propTypes = {
  realEstate: PropTypes.shape({
    name: PropTypes.string,
    address: PropTypes.shape({
      line1: PropTypes.string,
      line2: PropTypes.string,
      line3: PropTypes.string,
      locality: PropTypes.string,
      state: PropTypes.string,
      postcode: PropTypes.string,
    }),
    notes: PropTypes.string,
    purchaseDate: PropTypes.string,
    saleDate: PropTypes.string,
    valuations: PropTypes.arrayOf(
      PropTypes.shape({
        date: PropTypes.string,
        amount: PropTypes.number,
        note: PropTypes.string,
        type: PropTypes.oneOf([ 'none', 'purchase', 'sale' ]),
      })
    ),
  }),
  onSubmit: PropTypes.func,
}

export default RealEstateForm
