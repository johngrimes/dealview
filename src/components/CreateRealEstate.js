import React from 'react'
import { connect } from 'react-redux'
import { withRouter } from 'react-router'

import RealEstateForm from './RealEstateForm.js'
import Breadcrumbs from './Breadcrumbs.js'
import RealEstateActions from '../actions/realEstate.js'

import './styles/CreateRealEstate.css'

export class CreateRealEstate extends React.Component {
  constructor(props) {
    super(props)

    this.handleSubmit = this.handleSubmit.bind(this)
  }

  breadcrumbs() {
    return [
      { display: 'Portfolio', path: '/portfolio' },
      { display: 'Assets', path: '/portfolio/assets' },
      {
        display: 'New Real Estate Asset',
        path: '/portfolio/assets/real-estate/new',
      },
    ]
  }

  handleSubmit(realEstate) {
    this.props.dispatch(RealEstateActions.putRealEstate(realEstate))
    this.props.history.push('/portfolio/assets')
  }

  render() {
    return (
      <div className='create-real-estate'>
        <Breadcrumbs breadcrumbs={this.breadcrumbs()} />
        <RealEstateForm onSubmit={this.handleSubmit} />
      </div>
    )
  }
}

export default connect()(withRouter(CreateRealEstate))
