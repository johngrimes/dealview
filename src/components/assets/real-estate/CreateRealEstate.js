// @flow

import React from 'react'
import type { Children } from 'react'

import RealEstateForm from './RealEstateForm.js'

const CreateRealEstate = (props: { children: Children }) => {
  const { children } = props

  return (
    <div className="create-real-estate">
      <h1>Create Real Estate</h1>
      <RealEstateForm/>
      {children}
    </div>
  )
}

export default CreateRealEstate
