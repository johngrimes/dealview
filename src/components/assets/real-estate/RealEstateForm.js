// @flow

import React from 'react'
import _ from 'lodash'
import moment from 'moment'

import InputField from 'components/forms/InputField'
import HiddenField from 'components/forms/HiddenField'
import AddressField, { AddressErrorsDefaults } from 'components/forms/AddressField'
import TextAreaField from 'components/forms/TextAreaField'
import ValuationsInput from 'components/forms/ValuationsInput/ValuationsInput'
import DateField from 'components/forms/DateField'
import * as Validations from 'utils/FormValidation'
import { RealEstateDefaults } from 'types/assets/realEstate'
import { DateStorageFormat } from 'types/commonTypes'
import { updateValuationsWithPurchaseOrSale, updateValuationsWithPurchase,
  updateValuationsWithSale, purchaseFilter, saleFilter } from 'types/valuations'
import type { RealEstate } from 'types/assets/realEstate'
import type { Address } from 'types/commonTypes'
import type { AddressErrors } from 'components/forms/AddressField'
import type { Valuations } from 'types/valuations'
import type { FieldErrors } from 'utils/FormValidation'

import 'styles/forms.css'
import 'styles/buttons.css'
import 'styles/tables.css'
import 'components/assets/real-estate/RealEstateForm.css'

type RealEstateErrors = {
  +name: FieldErrors,
  +address: AddressErrors,
  +notes: FieldErrors,
  +purchaseDate: FieldErrors,
  +purchaseAmount: FieldErrors,
  +saleDate: FieldErrors,
  +saleAmount: FieldErrors,
  +valuations: FieldErrors,
}
const RealEstateErrorsDefaults = {
  name: [],
  address: AddressErrorsDefaults,
  notes: [],
  purchaseDate: [],
  purchaseAmount: [],
  saleDate: [],
  saleAmount: [],
  valuations: [],
}

type Props = {
  +realEstate?: RealEstate,
  +onSubmit?: (realEstate: RealEstate) => void
}

type State = {
  +realEstate: RealEstate,
  +purchase?: { date: string, amount: number },
  +sale?: { date: string, amount: number },
  +errors?: RealEstateErrors,
  +allErrorsShown: boolean,
  +focusedInput: string
}

class RealEstateForm extends React.Component {
  props: Props
  state: State
  handleChange: (fieldName: string, value: string) => void
  handleFocus: (fieldName: string) => void
  handleAddressChange: (address: Address) => void
  handleValuationsChange: (valuations: Valuations) => void
  handlePurchaseChange: (purchase: { +date?: string, +amount?: number }) => void
  handleSaleChange: (sale: { +date?: string, +amount?: number }) => void
  handleSubmit: (event: Event) => boolean

  constructor(props: Props) {
    super(props)
    this.state = {
      realEstate: RealEstateDefaults,
      allErrorsShown: false,
      focusedInput: 'name',
    }

    if (props.realEstate) {
      const realEstate = props.realEstate
      const valuations = props.realEstate.valuations
        ? props.realEstate.valuations
        : []
      this.state = {
        ...this.state,
        realEstate,
        purchase: _.pick(valuations.find(purchaseFilter), [ 'date', 'amount' ]),
        sale: _.pick(valuations.find(saleFilter), [ 'date', 'amount' ]),
        errors: RealEstateForm.validate(realEstate),
        allErrorsShown: true,
      }
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleFocus = this.handleFocus.bind(this)
    this.handleAddressChange = this.handleAddressChange.bind(this)
    this.handleValuationsChange = this.handleValuationsChange.bind(this)
    this.handlePurchaseChange = this.handlePurchaseChange.bind(this)
    this.handleSaleChange = this.handleSaleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange(fieldName: string, value: string): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, [fieldName]: value }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleFocus(fieldName: string): void {
    this.setState(() => ({ focusedInput: fieldName }))
  }

  handleAddressChange(address: Address): void {
    this.setState(prevState => {
      const realEstate = { ...prevState.realEstate, address }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handleValuationsChange(valuations: Valuations): void {
    this.setState(prevState => {
      let purchaseVal = valuations.find(v => v.type === 'purchase')
      let saleVal = valuations.find(v => v.type === 'sale')
      let updatedValuations = valuations
      if (purchaseVal) {
        purchaseVal = _.pick(purchaseVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithPurchase(valuations, purchaseVal)
      }
      if (saleVal) {
        saleVal = _.pick(saleVal, [ 'date', 'amount' ])
        updatedValuations = updateValuationsWithSale(valuations, saleVal)
      }
      const realEstate = { ...prevState.realEstate, valuations: updatedValuations }
      return {
        realEstate,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseOrSaleChange(purchaseOrSale: { +date?: string, +amount?: number }, type: 'purchase'|'sale'): void {
    this.setState(prevState => {
      const updatedPurchaseOrSale = { ...prevState[type], ...purchaseOrSale }
      const updatedValuations = updateValuationsWithPurchaseOrSale(prevState.realEstate.valuations, updatedPurchaseOrSale, type)
      const realEstate = {
        ...prevState.realEstate,
        valuations: updatedValuations,
        [ type === 'purchase' ? 'purchaseDate' : 'saleDate' ]: updatedPurchaseOrSale.date,
      }
      return {
        ...prevState,
        realEstate,
        [ type === 'purchase' ? 'purchase' : 'sale' ]: updatedPurchaseOrSale,
        errors: RealEstateForm.validate(realEstate),
      }
    })
  }

  handlePurchaseChange(purchase: { +date?: string, +amount?: number }): void {
    return this.handlePurchaseOrSaleChange(purchase, 'purchase')
  }

  handleSaleChange(sale: { +date?: string, +amount?: number }): void {
    return this.handlePurchaseOrSaleChange(sale, 'sale')
  }

  handleSubmit(event: Event): boolean {
    event.preventDefault()
    this.setState(
      () => ({ errors: RealEstateForm.validate(this.state.realEstate) }),
      () => {
        const { errors, realEstate } = this.state
        if (typeof errors !== 'undefined' && Validations.areErrorsPresent(errors)) {
          const firstErrorFieldName = RealEstateForm.findFirstErrorFieldName(errors)
          firstErrorFieldName
            ? this.setState(() => ({ allErrorsShown: true, focusedInput: firstErrorFieldName }))
            : this.setState(() => ({ allErrorsShown: true }))
        } else {
          if (this.props.onSubmit) { this.props.onSubmit(realEstate) }
        }
      }
    )
    return false
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.realEstate) {
      this.setState(() => ({ realEstate: nextProps.realEstate }))
    }
  }

  render() {
    const { realEstate, purchase, sale, allErrorsShown, focusedInput } = this.state
    const errors = this.state.errors === undefined ? {} : this.state.errors
    const idField = typeof realEstate.id === 'string'
      ? <HiddenField name='id' value={realEstate.id} />
      : null
    const minValuationDate = purchase && purchase.date
      ? moment(purchase.date, DateStorageFormat).add(1, 'd') : undefined
    const maxValuationDate = sale && sale.date
      ? moment(sale.date, DateStorageFormat).subtract(1, 'd') : undefined

    return (
      <form className='real-estate-form form form-aligned' onSubmit={this.handleSubmit}>
        <fieldset>
          <legend>General details</legend>

          {idField}
          <InputField name='name' type='text' label='Name' value={realEstate.name}
            errors={errors.name} forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleChange('name', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
          <AddressField name='address' address={realEstate.address} errors={errors.address}
            onChange={this.handleAddressChange}
            onFocus={this.handleFocus} focus={focusedInput} />
          <TextAreaField name='notes' label='Notes' value={realEstate.notes}
            onChange={value => this.handleChange('notes', value)}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Purchase</legend>
          <DateField name='purchaseDate' label='Purchase date'
            value={purchase ? purchase.date : undefined} errors={errors.purchaseDate}
            forceErrorDisplay={allErrorsShown}
            maxDate={maxValuationDate}
            onChange={value => this.handlePurchaseChange({ date: value })}
            onFocus={this.handleFocus} focus={focusedInput} />
          <InputField name='purchaseAmount' label='Purchase amount' type='number' min='1'
            value={purchase && purchase.amount ? purchase.amount.toString() : undefined}
            errors={errors.purchaseAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handlePurchaseChange({ amount: parseInt(value) })}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Sale</legend>
          <DateField name='saleDate' label='Sale date'
            value={sale ? sale.date : undefined} errors={errors.saleDate}
            forceErrorDisplay={allErrorsShown}
            minDate={minValuationDate}
            onChange={value => this.handleSaleChange({ date: value })}
            onFocus={this.handleFocus} focus={focusedInput} />
          <InputField name='saleAmount' label='Sale amount' type='number' min='1'
            value={sale && sale.amount ? sale.amount.toString() : undefined}
            errors={errors.saleAmount}
            forceErrorDisplay={allErrorsShown}
            onChange={value => this.handleSaleChange({ amount: parseInt(value) })}
            onFocus={this.handleFocus} focus={focusedInput} />
        </fieldset>
        <fieldset>
          <legend>Value</legend>
          <ValuationsInput name='valuations' valuations={realEstate.valuations}
            minDate={minValuationDate} maxDate={maxValuationDate}
            errors={errors.valuations} forceErrorDisplay={allErrorsShown}
            onChange={this.handleValuationsChange} onFocus={this.handleFocus} />
        </fieldset>
        <fieldset>
          <legend>Mortgage</legend>
          <button type='submit' className='button button-primary'>Submit</button>
        </fieldset>
      </form>
    )
  }

  static validate(realEstate: RealEstate): RealEstateErrors {
    const name = []
      .concat(Validations.required(realEstate.name))
      .concat(Validations.minLength(realEstate.name, 3))
      .concat(Validations.maxLength(realEstate.name, 100))
    const address = AddressField.validate(realEstate.address)
    const purchaseDate = []
      .concat(Validations.required(realEstate.purchaseDate))
    const valuations = []
      .concat(ValuationsInput.validate(realEstate.valuations))
      .concat(RealEstateForm.purchaseValuationPresent(realEstate))
      .concat(RealEstateForm.valuationsWithinPurchaseAndSale(realEstate))

    return {
      ...RealEstateErrorsDefaults,
      name,
      address,
      purchaseDate,
      valuations,
    }
  }

  static purchaseValuationPresent(realEstate: RealEstate): FieldErrors {
    return Validations.validate(
      realEstate,
      re => re.valuations.some(valuation => valuation.type === 'purchase'),
      'There must be a purchase valuation'
    )
  }

  static valuationsWithinPurchaseAndSale(realEstate: RealEstate): FieldErrors {
    const purchaseVal = realEstate.valuations.find(v => v.type === 'purchase')
    const saleVal = realEstate.valuations.find(v => v.type === 'sale')
    if (!purchaseVal) { return [] }

    return Validations.validate(
      realEstate,
      re => re.valuations.every(valuation =>
        [ 'purchase', 'sale' ].includes(valuation.type) || (saleVal
          ? valuation.date > purchaseVal.date && valuation.date < saleVal.date
          : valuation.date > purchaseVal.date)
      ),
      'Valuations must all be between the purchase and sale dates'
    )
  }

  static findFirstErrorFieldName(realEstateErrors: RealEstateErrors): string|null {
    return (function seek(realEstateErrors, parentComponentName) {
      let match = null
      _.some(realEstateErrors, (errors, fieldName) => {
        if (errors instanceof Array) {
          if (errors.length > 0) {
            match = parentComponentName ? parentComponentName + '-' + fieldName : fieldName
            return true
          } else return false
        } else if (errors instanceof Object) {
          match = seek(errors, fieldName)
          return (match !== null)
        }
      })
      return match
    })(realEstateErrors)
  }
}

export default RealEstateForm
