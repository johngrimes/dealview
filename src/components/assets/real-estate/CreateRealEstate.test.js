/* global expect */

import { CreateRealEstate } from './CreateRealEstate.js'

import React from 'react'
import { shallow } from 'enzyme'

import RealEstateForm from './RealEstateForm.js'
import { validRealEstate1 } from '../../../test/fixtures/realEstate.js'

describe('CreateRealEstate', () => {
  it('should render', () => {
    const wrapper = shallow(<CreateRealEstate />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should put Real Estate when handleSubmit is called', () => {
    const dispatch = jest.fn()
    const realEstateActions = require('../../../actions/realEstate.js')
    realEstateActions.putRealEstate = jest.fn()
    const wrapper = shallow(<CreateRealEstate dispatch={dispatch} />)
    const handleSubmit = wrapper.find(RealEstateForm).prop('onSubmit')
    handleSubmit(validRealEstate1)
    expect(dispatch).toHaveBeenCalled()
    expect(realEstateActions.putRealEstate).toHaveBeenCalledWith(validRealEstate1)
  })
})
