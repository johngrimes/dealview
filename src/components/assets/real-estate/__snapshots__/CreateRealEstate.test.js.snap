exports[`CreateRealEstate should render 1`] = `
<div
  className="create-real-estate">
  <Breadcrumbs
    breadcrumbs={
      Array [
        Object {
          "display": "Portfolio",
          "path": "/portfolio",
        },
        Object {
          "display": "Assets",
          "path": "/portfolio/assets",
        },
        Object {
          "display": "New Real Estate Asset",
          "path": "/portfolio/assets/real-estate/new",
        },
      ]
    } />
  <RealEstateForm
    onSubmit={[Function]} />
</div>
`;
